File: nodejs/src/ui/input.js
import { createInterface } from 'readline';
import { Branding } from './branding.js';
import { columnWidth, centerPad } from './chatbox.js';

const C = Branding.colors;
const MAX_POPUP = 9;

const visible = (s) => String(s ?? '').replace(/\x1b\[[0-9;]*m/g, '').length;

function padRight(text, width) {
  const t = String(text ?? '');
  return t + ' '.repeat(Math.max(0, width - visible(t)));
}

/**
 * Raw-mode bottom-anchored line editor. Draws a fixed input bar on the last
 * terminal row and, when the draft starts with "/", a popup box listing
 * matching slash commands above it. Falls back to plain line reading when
 * stdin is not a TTY (piped input, tests).
 */
export class LineEditor {
  constructor({
    prompt = '',
    commands = [],
    onLine = () => {},
    onExit = () => {},
    stream = process.stdin,
    write = (s) => process.stdout.write(s),
    dims = () => ({ columns: process.stdout.columns || 96, rows: process.stdout.rows || 24 }),
  } = {}) {
    this.prompt = prompt;
    this.commands = commands;
    this.onLine = onLine;
    this.onExit = onExit;
    this.stream = stream;
    this.write = write;
    this.dims = dims;

    this.buffer = [];
    this.cursor = 0;
    this.viewOffset = 0;
    this.history = [];
    this.histIndex = -1;
    this.popup = null; // { matches, selected } | null
    this.questionText = null;
    this.busy = false;
    this._pending = null;
    this._pendingResolve = null;
    this.pendingLines = [];
    this.started = false;
    this._fallback = null;
    this._lastLine = '';
  }

  _cols() {
    return Math.max(20, this.dims().columns || 96);
  }

  _rows() {
    return Math.max(8, this.dims().rows || 24);
  }

  get isTTY() {
    return !!(this.stream && this.stream.isTTY);
  }

  // Rows the editor reserves at the bottom of the terminal.
  reservedRows() {
    if (!this.isTTY) return 0;
    return this.popup ? this._popupH() : 1;
  }

  _popupH() {
    const n = this.popup ? Math.min(this.popup.matches.length, MAX_POPUP) : 0;
    const h = Math.max(1, n) + 2;
    return Math.min(h, Math.max(2, this._rows() - 1));
  }

  _text() {
    return this.buffer.join('');
  }

  start() {
    if (this.started) return this;
    this.started = true;
    if (this.isTTY) {
      this._raw = true;
      this.stream.setRawMode(true);
      this.stream.resume();
      this.stream.on('data', (chunk) => this._feed(String(chunk)));
      this._onResize = () => this.draw();
      process.stdout.on('resize', this._onResize);
    } else {
      this._fallback = createInterface({ input: this.stream, output: process.stdout, terminal: false });
      this._fallback.on('line', (line) => {
        if (this._pendingResolve) {
          const r = this._pendingResolve;
          this._pendingResolve = null;
          this._pending = null;
          r(line);
        } else if (line !== undefined) {
          this.pendingLines.push(line);
        }
      });
      this._fallback.on('close', () => {
        this._pendingResolve?.(null);
        this._pendingResolve = null;
        this._pending = null;
      });
      process.stdout.write(this.prompt);
    }
    this.draw();
    return this;
  }

  stop() {
    if (this._raw) {
      try {
        this.stream.setRawMode(false);
      } catch { /* noop */ }
      this.stream.removeAllListeners('data');
      this._raw = false;
    }
    if (this._onResize) {
      process.stdout.removeListener('resize', this._onResize);
      this._onResize = null;
    }
    this._fallback?.close();
    this._fallback = null;
  }

  // Resolves any pending read with null and restores the terminal.
  exit() {
    if (this._pendingResolve) {
      const r = this._pendingResolve;
      this._pendingResolve = null;
      this._pending = null;
      r(null);
    }
    this.stop();
  }

  readLine() {
    if (this.pendingLines.length) {
      const line = this.pendingLines.shift();
      this._recordHistory(line);
      return Promise.resolve(line);
    }
    if (this._pending) return this._pending;
    this._pending = new Promise((resolve) => {
      this._pendingResolve = resolve;
    });
    return this._pending;
  }

  question(msg) {
    this.questionText = String(msg ?? '');
    this.buffer = [];
    this.cursor = 0;
    this.popup = null;
    this.draw();
    return this.readLine();
  }

  setBusy(flag) {
    this.busy = !!flag;
  }

  // ----- internal state manipulation -----

  _recordHistory(line) {
    if (line && line !== this.history[this.history.length - 1]) {
      this.history.push(line);
      if (this.history.length > 200) this.history.shift();
    }
    this.histIndex = -1;
  }

  _resolve(line) {
    this._recordHistory(line);
    this.questionText = null;
    this.buffer = [];
    this.cursor = 0;
    this.viewOffset = 0;
    this.popup = null;
    if (this._pendingResolve) {
      const r = this._pendingResolve;
      this._pendingResolve = null;
      this._pending = null;
      r(line);
    } else if (line !== undefined) {
      this.pendingLines.push(line);
    }
  }

  _matches(text) {
    const token = text.startsWith('/') ? text.slice(1) : '';
    const base = token.trim().split(/ +/)[0] || '';
    return this.commands.filter((c) => c.name.startsWith(base));
  }

  _syncPopup() {
    const s = this._text();
    if (!s.startsWith('/')) {
      this.popup = null;
      return;
    }
    const words = s.slice(1).split(/ +/);
    const base = words[0] || '';
    if (words.length >= 2 && this.commands.some((c) => c.name === base)) {
      this.popup = null;
      return;
    }
    const matches = this._matches(s);
    const sel = this.popup ? Math.min(this.popup.selected, Math.max(0, matches.length - 1)) : -1;
    this.popup = { matches, selected: matches.length ? sel : -1 };
  }

  _acceptPopup(submit) {
    const { matches } = this.popup;
    const m = matches.length ? matches[this.popup.selected >= 0 ? this.popup.selected : 0] : null;
    if (!m) {
      this.popup = null;
      if (submit) this._resolve(this._text());
      return;
    }
    if (m.args && submit) {
      // Needs arguments: fill the command and keep editing.
      this.buffer = `/${m.name} `.split('');
      this.cursor = this.buffer.length;
      this.viewOffset = 0;
      this.popup = null;
    } else {
      this.popup = null;
      this._resolve(`/${m.name}`);
    }
  }

  _submit() {
    if (this.busy) {
      // Streaming: stash the line, deliver it right after the busy phase.
      const line = this._text();
      this._recordHistory(line);
      this.buffer = [];
      this.cursor = 0;
      this.popup = null;
      this.pendingLines.push(line);
      return;
    }
    if (this.popup) {
      this._acceptPopup(true);
      return;
    }
    this._resolve(this._text());
  }

  // ----- key handling -----

  _feed(chunk) {
    const code = [...String(chunk)];
    let i = 0;
    while (i < code.length) {
      const c = code[i].charCodeAt(0);
      if (c === 27) {
        const esc = this._escape(code, i);
        i = esc.next;
        this._key(esc.key);
      } else if (c === 13 || c === 10) {
        this._submit();
        i++;
      } else if (c === 3) {
        this._ctrlC();
        i++;
      } else if (c === 4) {
        this._ctrlD();
        i++;
      } else if (c === 9) {
        this._tab();
        i++;
      } else if (c === 127) {
        this._backspace();
        i++;
      } else if (c < 32) {
        i++;
      } else {
        this._insert(code[i]);
        i++;
      }
    }
    this.draw();
  }

  _escape(code, i) {
    if (code[i + 1] !== '[') return { next: i + 1, key: 'esc' };
    let j = i + 2;
    let param = '';
    while (j < code.length && /[0-9;]/.test(code[j])) {
      param += code[j];
      j++;
    }
    const fin = code[j] || '';
    const next = j + 1;
    if (fin === 'A') return { next, key: 'up' };
    if (fin === 'B') return { next, key: 'down' };
    if (fin === 'C') return { next, key: 'right' };
    if (fin === 'D') return { next, key: 'left' };
    if (fin === 'H' || param === '1' || param === '7') return { next, key: 'home' };
    if (fin === 'F' || param === '4' || param === '8') return { next, key: 'end' };
    if (param === '3') return { next, key: 'delete' };
    if (param === 'Z') return { next, key: 'shift-tab' };
    if (param === '1;5' && fin === 'C') return { next, key: 'ctrl-right' };
    if (param === '1;5' && fin === 'D') return { next, key: 'ctrl-left' };
    return { next, key: 'ignore' };
  }

  _insert(ch) {
    this.buffer.splice(this.cursor, 0, ch);
    this.cursor++;
    this._syncPopup();
  }

  _backspace() {
    if (this.cursor > 0) {
      this.buffer.splice(this.cursor - 1, 1);
      this.cursor--;
      this._syncPopup();
    }
  }

  _tab() {
    if (this.popup) {
      const { matches } = this.popup;
      if (matches.length) {
        this.popup.selected = Math.min(matches.length - 1, this.popup.selected + 1);
      }
      return;
    }
    if (this._text().startsWith('/')) {
      this._syncPopup();
      if (this.popup && this.popup.matches.length) this.popup.selected = Math.min(1, this.popup.matches.length - 1);
      return;
    }
    this.buffer.splice(this.cursor, 0, ' ', ' ');
    this.cursor += 2;
  }

  _ctrlC() {
    if (this.popup) {
      this.popup = null;
    } else {
      this.onExit();
    }
  }

  _ctrlD() {
    if (!this._text()) this.onExit();
  }

  _key(key) {
    switch (key) {
      case 'up':
        if (this.popup && this.popup.matches.length) {
          this.popup.selected = Math.max(0, this.popup.selected - 1);
        } else {
          this._hist(-1);
        }
        break;
      case 'down':
        if (this.popup && this.popup.matches.length) {
          this.popup.selected = Math.min(this.popup.matches.length - 1, this.popup.selected + 1);
        } else {
          this._hist(1);
        }
        break;
      case 'shift-tab':
        if (this.popup && this.popup.matches.length) {
          this.popup.selected = Math.max(0, this.popup.selected - 1);
        }
        break;
      case 'left':
        this.cursor = Math.max(0, this.cursor - 1);
        break;
      case 'right':
        this.cursor = Math.min(this.buffer.length, this.cursor + 1);
        break;
      case 'ctrl-left':
        this._wordJump(-1);
        break;
      case 'ctrl-right':
        this._wordJump(1);
        break;
      case 'home':
        this.cursor = 0;
        break;
      case 'end':
        this.cursor = this.buffer.length;
        break;
      case 'delete':
        if (this.cursor < this.buffer.length) this.buffer.splice(this.cursor, 1);
        this._syncPopup();
        break;
      case 'esc':
        this.popup = null;
        break;
      default:
        break;
    }
  }

  _wordJump(dir) {
    if (dir < 0) {
      while (this.cursor > 0 && this.buffer[this.cursor - 1] === ' ') this.cursor--;
      while (this.cursor > 0 && this.buffer[this.cursor - 1] !== ' ') this.cursor--;
    } else {
      while (this.cursor < this.buffer.length && this.buffer[this.cursor] === ' ') this.cursor++;
      while (this.cursor < this.buffer.length && this.buffer[this.cursor] !== ' ') this.cursor++;
    }
  }

  _hist(dir) {
    if (!this.history.length) return;
    if (this.histIndex === -1) {
      this._lastLine = this._text();
      this.histIndex = this.history.length;
    }
    this.histIndex = Math.min(this.history.length, Math.max(0, this.histIndex + dir));
    if (this.histIndex === this.history.length) {
      this.buffer = this._lastLine.split('');
    } else {
      this.buffer = this.history[this.histIndex].split('');
    }
    this.cursor = this.buffer.length;
  }

  // ----- drawing -----

  _inputPrefix() {
    if (this.questionText !== null) {
      return `  ${C.warning}${this.questionText} (y/N)${C.reset} ${C.dim}${C.bold}${C.primary}${Branding.USER_PROMPT}${C.reset} ${C.dim}›${C.reset} `;
    }
    return `${this.prompt} `;
  }

  draw() {
    if (!this.isTTY) return;
    const R = this._rows();
    const cols = this._cols();
    const col = columnWidth(cols);
    const pad = centerPad(col, cols);
    const prevH = this._lastPopupH || 1;
    const curH = this.popup ? this._popupH() : 1;
    const clearTop = Math.max(1, R - Math.max(prevH, curH) + 1);
    let out = '';
    for (let r = clearTop; r <= R; r++) out += `\x1b[${r};1H\x1b[2K`;
    this._lastPopupH = curH;

    if (this.popup) {
      const boxH = this._popupH();
      const listRows = Math.min(this.popup.matches.length, MAX_POPUP);
      const box = Math.max(8, col);
      const inner = Math.max(1, box - 8);
      const top = R - boxH + 1;

      const head = `  ${C.primary}╭─${C.reset} ${C.bold}${C.accent}Commands${C.reset}${C.primary} `;
      const fill = Math.max(0, box - visible(head) - 1);
      out += `\x1b[${top};1H\x1b[2K${pad}${head}${'─'.repeat(fill)}╮${C.reset}`;

      let li = top + 1;
      for (let i = 0; i < listRows; i++) {
        const m = this.popup.matches[i];
        const selected = i === this.popup.selected;
        const name = `${C.bold}${selected ? C.accent : C.primary}/${m.name}${C.reset}`;
        const args = m.args ? ` ${C.dim}${m.args}${C.reset}` : '';
        const desc = m.desc ? `  ${C.dim}${m.desc}${C.reset}` : '';
        const body = ` ${' '.repeat(selected ? 0 : 3)}${selected ? '›' : '·'} ${name}${args}${desc}`;
        out += `\x1b[${li};1H\x1b[2K${pad}  ${C.primary}│${C.reset} ${selected ? C.accent + C.bold : ''}${padRight(body, inner)}${selected ? C.reset : ''} ${C.primary}│${C.reset}`;
        li++;
      }
      if (listRows === 0) {
        out += `\x1b[${li};1H\x1b[2K${pad}  ${C.primary}│${C.reset} ${C.dim}${padRight(' (no matching commands)', inner)}${C.reset} ${C.primary}│${C.reset}`;
        li++;
      }
      const hint = listRows ? '↑↓  Tab pick   Enter run   Esc close' : 'Esc close';
      const foot = `  ${C.primary}╰${'─'.repeat(Math.max(0, box - 6))}╯${C.reset}`;
      out += `\x1b[${li};1H\x1b[2K${pad}${foot}${C.dim} ${hint}${C.reset}`;
    }

    // Input bar on the last row.
    const max = Math.max(1, cols - visible(this._inputPrefix()) - 1);
    let off = this.viewOffset;
    if (this.cursor < off) off = this.cursor;
    if (this.cursor > off + max - 1) off = this.cursor - max + 1;
    this.viewOffset = off;

    const shown = this._text().slice(off, off + max);
    out += `\x1b[${R};1H\x1b[2K${pad}${this._inputPrefix()}${shown}`;
    const cursorCol = visible(pad + this._inputPrefix()) + (this.cursor - off) + 1;
    out += `\x1b[${R};${Math.min(cursorCol, cols)}H`;
    this.write(out);
  }
}

File: python/vierrataleai/ui/input.py
import os
import signal
import sys
import threading
import termios
import tty

from .branding import Branding
from .chatbox import C, _visible, _pad_right

MAX_POPUP = 9


def _column_width(term_width):
    w = term_width if term_width and term_width > 44 else 96
    return max(40, min(100, w - 8))


def _center_pad(col, term_width):
    w = term_width if term_width and term_width > 44 else 96
    return " " * max(0, (w - col) // 2)


class LineEditor:
    """Raw-mode bottom-anchored line editor.

    Draws a fixed input bar on the last terminal row and, when the draft starts
    with "/", a popup box listing matching slash commands above it. Falls back
    to plain line reading when stdin is not a TTY (piped input, tests).
    """

    def __init__(self, prompt="", commands=None, on_exit=None, stream=None, write=None, dims=None):
        self.prompt = prompt
        self.commands = commands or []
        self.on_exit = on_exit
        self.stream = stream if stream is not None else sys.stdin
        self.write = write if write is not None else (lambda s: (sys.stdout.write(s), sys.stdout.flush()))
        self.dims = dims

        self.buffer = []
        self.cursor = 0
        self.view_offset = 0
        self.history = []
        self.hist_index = -1
        self.last_line = ""
        self.popup = None
        self.question_text = None
        self.busy = False
        self.pending_lines = []
        self.started = False
        self._exit_requested = False
        self._lock = threading.Lock()
        self._saved_flags = None
        self._old_winch = None

    # ----- geometry -----

    def _cols(self):
        if self.dims:
            return max(20, self.dims()[0] or 96)
        try:
            import shutil
            return max(20, shutil.get_terminal_size().columns)
        except Exception:
            return 96

    def _rows(self):
        if self.dims:
            return max(8, self.dims()[1] or 24)
        try:
            import shutil
            return max(8, shutil.get_terminal_size().lines)
        except Exception:
            return 24

    @property
    def is_tty(self):
        try:
            return bool(self.stream.isatty())
        except Exception:
            return False

    def reserved_rows(self):
        if not self.is_tty:
            return 0
        return self._popup_h() if self.popup else 1

    def _popup_h(self):
        n = min(len(self.popup["matches"]), MAX_POPUP) if self.popup else 0
        return min(max(1, n) + 2, max(2, self._rows() - 1))

    def _text(self):
        return "".join(self.buffer)

    # ----- lifecycle -----

    def start(self):
        if self.started:
            return self
        self.started = True
        if self.is_tty:
            fd = self.stream.fileno()
            self._saved_flags = termios.tcgetattr(fd)
            tty.setraw(fd)
            self._old_winch = signal.signal(signal.SIGWINCH, lambda *_: self.draw())
        else:
            try:
                self.write(self.prompt)
            except Exception:
                pass
        self.draw()
        return self

    def stop(self):
        if self._saved_flags is not None:
            try:
                termios.tcsetattr(self.stream.fileno(), termios.TCSADRAIN, self._saved_flags)
            except Exception:
                pass
            self._saved_flags = None
        if self._old_winch is not None:
            try:
                signal.signal(signal.SIGWINCH, self._old_winch)
            except Exception:
                pass
            self._old_winch = None

    def exit(self):
        self._exit_requested = True
        self._submit_if_pending()
        self.stop()

    def _submit_if_pending(self):
        if self._text():
            self.buffer = []
            self.cursor = 0

    # ----- public reading API (blocking; run from an executor thread) -----

    def read_line(self):
        if self._exit_requested:
            return None
        if self.pending_lines:
            line = self.pending_lines.pop(0)
            self._record_history(line)
            return line
        if self.is_tty:
            while not self._exit_requested:
                key = self._read_key()
                if key is None:
                    self._feed_key("ctrl-d")
                else:
                    self._feed_char(key)
                if self.pending_lines:
                    line = self.pending_lines.pop(0)
                    self._record_history(line)
                    return line
            return None
        try:
            raw = self.stream.readline()
        except (KeyboardInterrupt, EOFError):
            return None
        line = raw.rstrip("\r\n")
        self._record_history(line)
        return line

    def question(self, text):
        self.question_text = str(text or "")
        self.buffer = []
        self.cursor = 0
        self.popup = None
        self.draw()
        return self.read_line()

    def set_busy(self, flag):
        self.busy = bool(flag)

    def _feed_key(self, key):
        if key == "ctrl-d":
            self._ctrl_d()

    # ----- internal state -----

    def _record_history(self, line):
        if line and (not self.history or line != self.history[-1]):
            self.history.append(line)
            if len(self.history) > 200:
                self.history.pop(0)
        self.hist_index = -1

    def _push(self, line):
        self._record_history(line)
        self.pending_lines.append(line)

    def _submit(self):
        line = self._text()
        self.question_text = None
        if self.busy:
            self._push(line)
            self.buffer = []
            self.cursor = 0
            self.popup = None
            self.draw()
            return
        if self.popup:
            self._accept_popup()
            self.draw()
            return
        self._push(line)
        self.buffer = []
        self.cursor = 0
        self.popup = None
        self.draw()

    def _matches(self, text):
        if not text.startswith("/"):
            return []
        base = text[1:].strip().split()[0] if len(text) > 1 else ""
        return [c for c in self.commands if c["name"].startswith(base)]

    def _sync_popup(self):
        s = self._text()
        if not s.startswith("/"):
            self.popup = None
            return
        words = s[1:].split(" ")
        base = words[0] or ""
        if len(words) >= 2 and any(c["name"] == base for c in self.commands):
            self.popup = None
            return
        matches = self._matches(s)
        if not matches:
            self.popup = None
            return
        sel = min(self.popup["selected"], len(matches) - 1) if self.popup and self.popup["selected"] >= 0 else -1
        self.popup = {"matches": matches, "selected": sel}

    def _accept_popup(self):
        m = self.popup["matches"][self.popup["selected"] if self.popup["selected"] >= 0 else 0] if self.popup and self.popup["matches"] else None
        if not m:
            self.popup = None
            self._push(self._text())
            self.buffer = []
            self.cursor = 0
            self.draw()
            return
        if m.get("args"):
            self.buffer = list("/%s " % m["name"])
            self.cursor = len(self.buffer)
            self.view_offset = 0
            self.popup = None
            self.draw()
        else:
            self.popup = None
            self._push("/%s" % m["name"])
            self.buffer = []
            self.cursor = 0
            self.draw()

    def _tab(self):
        if self.popup:
            matches = self.popup["matches"]
            if matches:
                self.popup["selected"] = min(len(matches) - 1, self.popup["selected"] + 1)
            return
        if self._text().startswith("/"):
            self._sync_popup()
            return
        self.buffer[self.cursor:self.cursor] = [" ", " "]
        self.cursor += 2

    def _ctrl_c(self):
        if self.popup:
            self.popup = None
        elif self.on_exit:
            self._exit_requested = True
            self.on_exit()
            self.stop()
        else:
            self._exit_requested = True
            self.stop()

    def _ctrl_d(self):
        if not self._text():
            self._exit_requested = True
            if self.on_exit:
                self.on_exit()
                self.stop()

    def _backspace(self):
        if self.cursor > 0:
            del self.buffer[self.cursor - 1]
            self.cursor -= 1
            self._sync_popup()

    def _insert(self, ch):
        self.buffer.insert(self.cursor, ch)
        self.cursor += 1
        self._sync_popup()

    def _key(self, key):
        if key == "up":
            if self.popup and self.popup["matches"]:
                self.popup["selected"] = max(0, self.popup["selected"] - 1)
            else:
                self._hist(-1)
        elif key == "down":
            if self.popup and self.popup["matches"]:
                self.popup["selected"] = min(len(self.popup["matches"]) - 1, self.popup["selected"] + 1)
            else:
                self._hist(1)
        elif key == "shift-tab":
            if self.popup and self.popup["matches"]:
                self.popup["selected"] = max(0, self.popup["selected"] - 1)
        elif key == "left":
            self.cursor = max(0, self.cursor - 1)
        elif key == "right":
            self.cursor = min(len(self.buffer), self.cursor + 1)
        elif key == "ctrl-left":
            self._word_jump(-1)
        elif key == "ctrl-right":
            self._word_jump(1)
        elif key == "home":
            self.cursor = 0
        elif key == "end":
            self.cursor = len(self.buffer)
        elif key == "delete":
            if self.cursor < len(self.buffer):
                del self.buffer[self.cursor]
            self._sync_popup()
        elif key == "esc":
            self.popup = None

    def _word_jump(self, direction):
        if direction < 0:
            while self.cursor > 0 and self.buffer[self.cursor - 1] == " ":
                self.cursor -= 1
            while self.cursor > 0 and self.buffer[self.cursor - 1] != " ":
                self.cursor -= 1
        else:
            while self.cursor < len(self.buffer) and self.buffer[self.cursor] == " ":
                self.cursor += 1
            while self.cursor < len(self.buffer) and self.buffer[self.cursor] != " ":
                self.cursor += 1

    def _hist(self, direction):
        if not self.history:
            return
        if self.hist_index == -1:
            self.last_line = self._text()
            self.hist_index = len(self.history)
        self.hist_index = min(len(self.history), max(0, self.hist_index + direction))
        if self.hist_index == len(self.history):
            self.buffer = list(self.last_line)
        else:
            self.buffer = list(self.history[self.hist_index])
        self.cursor = len(self.buffer)

    # ----- input consumption -----

    def _feed(self, raw):
        for ch in raw:
            self._feed_char(ch)

    def _feed_char(self, ch):
        if not ch:
            return
        if ch == "\r" or ch == "\n":
            self._submit()
        elif ch == "\t":
            self._tab()
        elif ch == "\x03":
            self._ctrl_c()
        elif ch == "\x04":
            self._ctrl_d()
        elif ch == "\x7f":
            self._backspace()
        elif ch == "\x1b[A":
            self._key("up")
        elif ch == "\x1b[B":
            self._key("down")
        elif ch == "\x1b[C":
            self._key("right")
        elif ch == "\x1b[D":
            self._key("left")
        elif ch == "\x1b[H":
            self._key("home")
        elif ch == "\x1b[F":
            self._key("end")
        elif ch == "\x1b[3~":
            self._key("delete")
        elif ch == "\x1b[Z":
            self._key("shift-tab")
        elif ch == "\x1b[1;5C":
            self._key("ctrl-right")
        elif ch == "\x1b[1;5D":
            self._key("ctrl-left")
        elif ch == "\x1b":
            self._key("esc")
        else:
            self._insert(ch)
        self.draw()

    # ----- key parsing (raw terminal) -----

    def _read_key(self):
        """Reads one decoded key/character from the raw terminal (blocking)."""
        fd = self.stream.fileno()
        b = os.read(fd, 1)
        if not b:
            return None
        byte = b[0]
        if byte == 0x1B:
            return self._read_escape(fd)
        if byte < 0x20 or byte == 0x7F:
            return chr(byte)
        if byte < 0x80:
            return chr(byte)
        if byte < 0xC0:
            return chr(byte)
        if byte < 0xE0:
            n = 2
        elif byte < 0xF0:
            n = 3
        else:
            n = 4
        data = b
        for _ in range(n - 1):
            extra = os.read(fd, 1)
            if not extra:
                break
            data += extra
        try:
            return data.decode("utf-8")
        except Exception:
            return None

    def _read_escape(self, fd):
        b2 = os.read(fd, 1)
        if not b2:
            return "\x1b"
        if b2 != b"[":
            return "\x1b"
        param = b""
        while True:
            c = os.read(fd, 1)
            if not c:
                break
            if c in b"0123456789;":
                param += c
            else:
                fin = c
                break
        try:
            fin = fin.decode("ascii")
        except Exception:
            fin = ""
        if fin == "A":
            return "\x1b[A"
        if fin == "B":
            return "\x1b[B"
        if fin == "C":
            return "\x1b[C"
        if fin == "D":
            return "\x1b[D"
        if fin == "H" or param in (b"1", b"7"):
            return "\x1b[H"
        if fin == "F" or param in (b"4", b"8"):
            return "\x1b[F"
        if param == b"3":
            return "\x1b[3~"
        if param == b"Z":
            return "\x1b[Z"
        if param == b"1;5" and fin == "C":
            return "\x1b[1;5C"
        if param == b"1;5" and fin == "D":
            return "\x1b[1;5D"
        return "\x1b"

    # ----- drawing -----

    def _input_prefix(self):
        if self.question_text is not None:
            return (
                "  \033[33m%s (y/N)\033[0m \033[1m\033[38;2;124;58;237m%s\033[0m \033[2m›\033[0m "
                % (self.question_text, Branding.USER_PROMPT)
            )
        return "%s " % self.prompt

    def draw(self):
        if not self.is_tty:
            return
        with self._lock:
            R = self._rows()
            cols = self._cols()
            col = _column_width(cols)
            pad = _center_pad(col, cols)

            prev_h = getattr(self, "_last_popup_h", 1)
            cur_h = self._popup_h() if self.popup else 1
            clear_top = max(1, R - max(prev_h, cur_h) + 1)
            out = []
            for r in range(clear_top, R + 1):
                out.append("\x1b[%d;1H\x1b[2K" % r)
            self._last_popup_h = cur_h

            if self.popup:
                box_h = self._popup_h()
                list_rows = min(len(self.popup["matches"]), MAX_POPUP)
                box = max(8, col)
                inner = max(1, box - 8)
                top = R - box_h + 1
                head = "  \033[38;2;124;58;237m╭─\033[0m \033[1m\033[38;2;6;182;212mCommands\033[0m\033[38;2;124;58;237m "
                fill = max(0, box - _visible(head) - 1)
                out.append("\x1b[%d;1H%s%s%s%s╮\033[0m" % (top, pad, head, "─" * fill, (" " * 0)))
                li = top + 1
                for i in range(list_rows):
                    m = self.popup["matches"][i]
                    selected = i == self.popup["selected"]
                    name = "\033[1m\033[38;2;6;182;212m/%s\033[0m" % m["name"] if selected else "\033[1m\033[38;2;124;58;237m/%s\033[0m" % m["name"]
                    args = " \033[2m%s\033[0m" % m.get("args", "") if m.get("args") else ""
                    desc = "  \033[2m%s\033[0m" % m.get("desc", "") if m.get("desc") else ""
                    body = " %s %s%s%s" % ("›" if selected else "·", name, args, desc)
                    style = "\033[38;2;6;182;212m\033[1m" if selected else ""
                    out.append("\x1b[%d;1H\x1b[2K%s  \033[38;2;124;58;237m│\033[0m %s%s%s \033[38;2;124;58;237m│\033[0m" % (li, pad, style, _pad_right(body, inner), "\033[0m" if selected else ""))
                    li += 1
                hint = "\033[2m ↑↓  Tab pick   Enter run   Esc close\033[0m" if list_rows else "\033[2m Esc close\033[0m"
                foot = "  \033[38;2;124;58;237m╰%s╯\033[0m" % ("─" * max(0, box - 6))
                out.append("\x1b[%d;1H\x1b[2K%s%s%s" % (li, pad, foot, hint))
                o = self._text()
                prefix_len = _visible(pad + self._input_prefix())
                max_vis = max(1, cols - prefix_len - 1)
                off = self.view_offset
                if self.cursor < off:
                    off = self.cursor
                if self.cursor > off + max_vis - 1:
                    off = self.cursor - max_vis + 1
                self.view_offset = off
                shown = o[off:off + max_vis]
                out.append("\x1b[%d;1H\x1b[2K%s%s%s" % (R, pad, self._input_prefix(), shown))
                cursor_col = min(prefix_len + (self.cursor - off) + 1, cols)
                out.append("\x1b[%d;%dH" % (R, cursor_col))
                self.write("".join(out))
            else:
                o = self._text()
                prefix_len = _visible(pad + self._input_prefix())
                max_vis = max(1, cols - prefix_len - 1)
                off = self.view_offset
                if self.cursor < off:
                    off = self.cursor
                if self.cursor > off + max_vis - 1:
                    off = self.cursor - max_vis + 1
                self.view_offset = off
                shown = o[off:off + max_vis]
                out.append("\x1b[%d;1H\x1b[2K%s%s%s" % (R, pad, self._input_prefix(), shown))
                cursor_col = min(prefix_len + (self.cursor - off) + 1, cols)
                out.append("\x1b[%d;%dH" % (R, cursor_col))
                self.write("".join(out))
